testsuite/smoke-test/src/consensus_key_rotation.rs

```diff
@@ -0,0 +1,192 @@
+            let mut validator_override_config =
+                OverrideNodeConfig::load_config(config_path.clone()).unwrap();
+            validator_override_config
+                .override_config_mut()
+                .consensus
+                .safety_rules
+                .initial_safety_rules_config
+                .overriding_identity_blob_paths_mut()
+                .push(new_identity_path);
+            validator_override_config.save_config(config_path).unwrap();
+
+            info!("Restarting the node.");
+            validator.start().unwrap();
+            info!("Let it bake for 5 secs.");
+            tokio::time::sleep(Duration::from_secs(5)).await;
+            (operator_addr, new_pk, pop, operator_idx)
+        } else {
+            unreachable!()
+        };
+
+    info!("Update on-chain. Retry is needed in case randomness is enabled.");
+    swarm
+        .chain_info()
+        .into_aptos_public_info()
+        .mint(operator_addr, 99999999999)
+        .await
+        .unwrap();
+    let mut attempts = 10;
+    while attempts > 0 {
+        attempts -= 1;
+        let gas_options = GasOptions {
+            gas_unit_price: Some(100),
+            max_gas: Some(200000),
+            expiration_secs: 60,
+        };
+        let update_result = cli
+            .update_consensus_key(
+                operator_idx,
+                None,
+                new_pk.clone(),
+                pop.clone(),
+                Some(gas_options),
+            )
+            .await;
+        info!("update_result={:?}", update_result);
+        if let Ok(txn_smry) = update_result {
+            if txn_smry.success == Some(true) {
+                break;
+            }
+        }
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
+
+    assert!(attempts >= 1);
+
+    info!("Wait for epoch 5.");
+    wait_until_epoch(
+        &rest_client,
+        5,
+        Duration::from_secs(epoch_duration_secs * 2),
+    )
+    .await
+    .unwrap();
+    info!("Epoch 5 arrived.");
+
+    info!("All nodes should be alive.");
+    let liveness_check_result = swarm
+        .liveness_check(Instant::now().add(Duration::from_secs(30)))
+        .await;
+    assert!(liveness_check_result.is_ok());
+
+    info!("On-chain pk should be updated.");
+    let validator_set = get_on_chain_resource::<ValidatorSet>(&rest_client).await;
+    let verifier = ValidatorVerifier::from(&validator_set);
+    assert_eq!(new_pk, verifier.get_public_key(&operator_addr).unwrap());
+}
+
+async fn wait_until_epoch(
+    rest_cli: &Client,
+    target_epoch: u64,
+    time_limit: Duration,
+) -> anyhow::Result<()> {
+    let timer = Instant::now();
+    while timer.elapsed() < time_limit {
+        let c = get_on_chain_resource::<ConfigurationResource>(rest_cli).await;
+        if c.epoch() >= target_epoch {
+            return Ok(());
+        }
+        tokio::time::sleep(Duration::from_secs(1)).await;
+    }
+    bail!("");
```

Comment:
nit: maybe put a real error message here? 😄

Reply:
oops, i forgot auto-merge is on...